File: 55-jump-game/solution.cpp
#include <vector>
#include <algorithm>

class Solution
{
public:
  bool canJump(std::vector<int> &nums)
  {
    int max_index = -1;
    for (int i = 0; i < nums.size(); ++i)
    {
      if (i > max_index && max_index != -1)
      {
        return false;
      }

      if (i + nums[i] >= nums.size() - 1)
      {
        return true;
      }

      max_index = std::max(max_index, i + nums[i]);
    }

    return false;
  }
};

File: 55-jump-game/solution.py
from typing import List


class Solution:
  def canJump(self, nums: List[int]) -> bool:
    max_index = -1
    for i, num in enumerate(nums):
      if i > max_index and max_index != -1:
        return False

      if i + num >= len(nums) - 1:
        return True

      max_index = max(max_index, i + num)

    return False


s = Solution()
print(s.canJump([3, 2, 1, 0, 4]))
